server/aesdsocket.c

```diff
@@ -0,0 +1,371 @@
+
+    		
+    		signal(SIGCHLD, SIG_IGN);
+    		signal(SIGHUP, SIG_IGN);
+
+    		  
+
+   		 if(chdir("/")==-1)
+   		 {
+   		 	perror("chdir unsuccesful");
+   		 	close(ret4);
+   		 	exit(EXIT_FAILURE);
+   		 	}
+   		 	
+  		 int fd1;
+  	 
+  	 	 fd1 = open ("/dev/null", O_RDWR);
+  		dup2(fd1, STDIN_FILENO) ;           		
+        	dup2(fd1, STDOUT_FILENO) ;
+        	dup2(fd1, STDERR_FILENO) ;
+            	close(STDIN_FILENO);
+            	close(STDOUT_FILENO);
+            	close(STDERR_FILENO);
+            }
+  	 }
+	
+  	
+  	
+  	 ret3= listen(ret, 10);
+  	  if(ret3 == -1)
+  	 {
+  	 	perror("socket listen failed");
+  	 	close(ret4);
+  	 	close(ret);
+  	 	return false;
+  	 	//exit(1);
+  	 }
+  	
+  	
+  	size1= sizeof(sockaddr1);
+  
+	 
+    int temp_size=buf_size;
+    int total_buffer=0;
+    
+   	rec_buf = (char *)malloc(sizeof(char) * buf_size);
+  	read_buf = (char *)malloc(sizeof(char) * buf_size);
+  	if((rec_buf==NULL)| (read_buf==NULL))	
+  	{
+  		perror("malloc failed for buffers");
+  		close(ret4);
+  	 	close(ret);
+  	 			
+  	
+  	}
+  	 
+     fd= open("/var/tmp/aesdsocketdata.txt", O_CREAT | O_RDWR | O_APPEND , 0644);
+  	 if(fd == -1)
+  	 {
+  	  	perror("File create and open unsuccessful\n");
+  	  	close(ret4);
+  	  	close(ret);
+  	 	free(rec_buf); 
+  	 	free(read_buf); 		 	
+  		close(fd);
+  	  	return false;
+  	 }
+ 	 
+  while(signal_bool==0)
+ 	{
+ 	
+  	 
+  	 	
+  	 
+  	 
+  	 
+  	 
+  	 int len=0, len2, len3=0;  	 
+  	 int nr,nr1;   	
+  	  	
+  	  	
+  	 ret4= accept(ret, &sockaddr1, &size1 );
+  	 if(ret4 == -1)
+  	 {
+  	 	perror("socket accept failed");
+  	 	close(ret4);
+  	 	close(ret);
+  	 	free(read_buf); 
+  	 	close(fd);
+  	 	free(rec_buf); 
+  	 	return false;
+  	 	
+  	 	
+  	 }
+  	 else 
+  	 {
+  	 	struct sockaddr_in *addr_in = (struct sockaddr_in *)&sockaddr1;
+  	 	ip = inet_ntoa(addr_in->sin_addr);
+  	 	syslog(LOG_USER, "Accepted connection from %s",ip);
+  	 	
+  	 }
+  	 
+  	 
+  	  
+  	 
+  	 
+  	
+  	int total=0;
+  	
+  		do
+  		{
+  		
+  		 len= recv(ret4, rec_buf+total, buf_size, 0);
+  	 	 if(len == -1)
+  	 	{
+  	 		perror("receive failed");
+  	 		close(ret4);
+  	 		close(ret);
+  	 		free(rec_buf);  
+  	 		free(read_buf); 		 	
+  			close(fd);
+  	 		return false;
+  	 		
+  	 	} 
+  	 	 
+  	 	total+=(len);
+  	 	temp_size+= (buf_size);
+  	 	
+  	 	rec_buf=realloc(rec_buf, sizeof(char)*temp_size);
+  	 	
+  	 	
+  	 	
+  	 	}while(strchr(rec_buf, '\n') == NULL);	
+  	 	
+  	 	total_buffer+=total;
+  	 	
+  	 	rec_buf[total]='\0';
+  	 	
+  	 		nr=write(fd, rec_buf, total);
+			if(nr == -1) 
+			{
+				perror("File write unsuccessful\n");
+				close(ret4);
+				close(ret);
+  	 			free(rec_buf); 
+  	 			free(read_buf);  		 	
+  				close(fd);
+				return false;
+			}
+  	 	
+  	 	
+  	 	
+		
+		
+		
+		syslog(LOG_USER, " received rec_buf = %s", rec_buf);
+	
+		
+
+	
+  	 
+  	 
+  	 read_buf= realloc(read_buf, sizeof(char)*total_buffer);
```

Comment:
@chaisuresh what assumption is made here about the size of RAM heap relative to the total size of all packets sent?